alternator/ttl.cc

```diff
@@ -113,4 +142,437 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    std::optional<dht::partition_range> next_partition_range() {
+        std::optional<dht::token_range> ret = next();
+        if (ret) {
+            return dht::to_partition_range(*ret);
+        } else {
+            return std::nullopt;
+        }
+    }
+};
+
+// Scan data in a list of token ranges in one table, looking for expired
+// items and deleting them.
+// Because of issue #9167, partition_ranges must have a single partition
+// for this code to work correctly.
+// FIXME: The long list of parameters below is ugly. This function should
+// calculate what it needs, or we can have an object with many of these
+// values pre-calculated in one function and then used in this one.
+static future<> scan_table_ranges(
+        service::storage_proxy& proxy,
+        schema_ptr s,
+        dht::partition_range_vector&& partition_ranges,
+        const bytes& column_name,
+        const std::optional<std::string>& member,
+        shared_ptr<cql3::selection::selection> selection,
+        service::query_state& query_state,
+        cql3::query_options& query_options,
+        lw_shared_ptr<query::read_command> command,
+        abort_source& abort_source)
+{
+    assert (partition_ranges.size() == 1); // otherwise issue #9167 will cause incorrect results.
+    auto p = service::pager::query_pagers::pager(s, selection, query_state, query_options, command, std::move(partition_ranges), nullptr);
+    while (!p->is_exhausted()) {
+        if (abort_source.abort_requested()) {
+            co_return;
+        }
+        // We don't to limit page size in number of rows because there is a
+        // builtin limit of the page's size in bytes. Setting this limit to 1
+        // is useful for debugging the paging code with moderate-size data.
+        uint32_t limit = std::numeric_limits<uint32_t>::max();
+        // FIXME: which timeout?
+        // FIXME: if read times out, need to retry it.
+        std::unique_ptr<cql3::result_set> rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
+        auto rows = rs->rows();
+        auto meta = rs->get_metadata().get_names();
+        std::optional<unsigned> expiration_column;
+        std::vector<unsigned> pk_columns;
+        std::vector<unsigned> ck_columns;
+        for (unsigned i = 0; i < meta.size(); i++) {
+            const cql3::column_specification& col = *meta[i];
+            if (col.name->name() == column_name) {
+                expiration_column = i;
+                break;
+            }
+        }
+        if (!expiration_column) {
+            continue;
+        }
+        // Find the key columns, which we will need for extracting the key
+        // the item to delete. FIXME: I feel there must be a much better
+        // way to look at results :-( Don't we have a scan API which will
+        // generate items with separately marked keys, instead of with this
+        // complicated result_set interface?
```

Comment:
I think that the key columns appear in order, so you can use the first `n` for partition key, and then the next `m` for clustering key, with `n` and `m` being corresponding key sizes - assuming that you explicitly ask for all key values, but here you do. But I guess that your code is safer with regard to some corner cases I haven't thought of

Reply:
I'll take a look at if this idea can work - although this can also be a followup, because it's an optimization (and also listed as a FIXME).

I'm not happy at all about all the transformation that this code does - basically we read a key from the sstable as a contiguous serialized key, and then we open it up to that fancy "result set" API and then we need to wrap it back up to be a key. Everything would have been much more efficient if the key was just kept unparsed. But I think we can live with this inefficiency for now. It took me a long time to get this API to work, I don't want to start looking (just yet) for alternative APIs.